src/FromFile.jl

```diff
@@ -16,43 +13,70 @@ function from_m(m::Module, path::String, ex::Expr)
     path = abspath(normpath(path))
     loading = Expr(ex.head)
     root = root_module(m)
-
-    if root === Main
-        toplevel = __toplevel__
-    else
-        toplevel_symbol = Symbol("#__toplevel__#")
-        if isdefined(root, toplevel_symbol) # package
-            toplevel = getfield(root, toplevel_symbol)
-        else
-            toplevel = Base.eval(root, :(baremodule $toplevel_symbol; using Base; end))
-        end
-    end
-    
-    file_module = load_module(toplevel, root, path)
+    file_module = load_module(root, path)
 
     for each in ex.args
         each isa Expr || continue
 
         if each.head === :(:) # using/import A: a, b, c
             m_name = each.args[1].args[1]
             m_name === :(.) && error("cannot load relative module from file")
-            push!(loading.args, Expr(:(:), Expr(:., fullname(file_module)..., m_name), each.args[2:end]...) )
+            if root === Main
+                push!(loading.args, Expr(:(:), Expr(:., :., fullname(file_module)..., m_name), each.args[2:end]...) )
+            else
+                push!(loading.args, Expr(:(:), Expr(:., fullname(file_module)..., m_name), each.args[2:end]...) )
+            end
         elseif each.head === :(.) # using/import A, B, C
             m_name = each.args[1] # module name
             m_name === :(.) && error("cannot load relative module from file")
-            push!(loading.args, Expr(:., fullname(file_module)..., m_name))
+
+            if root === Main
+                push!(loading.args, Expr(:., :., fullname(file_module)..., m_name))
+            else
+                push!(loading.args, Expr(:., fullname(file_module)..., m_name))
+            end
         else
             error("invalid syntax $ex")
         end
     end
+
+    if root === Main && !isdefined(Main, Symbol(path))
+        return quote
+            const $(Symbol(path)) = $file_module
+            $loading
+        end
+    end
     return loading
 end
 
-function load_module(toplevel::Module, root::Module, path::String)
+function load_module(root::Module, path::String)
     if root === Main
-        file_module_sym = Symbol(path)
+        return load_module_from_main(path)
     else
-        file_module_sym = Symbol(relpath(path, pathof(root)))
+        return load_module_from_package(root, path)
+    end
+end
+
+function load_module_from_main(path)
+    file_module_sym = Symbol(path)
+    # always reload file when it's in interactive mode and in Main
+    if haskey(loaded_path, path) && !isinteractive()
+        return loaded_path[path]
+    else
+        file_module = Base.eval(Base.__toplevel__, :(module $(file_module_sym) end))
+        Base.include(file_module, path)
+        loaded_path[path] = file_module
```

Comment:
I'm nots ure `loaded_path` is needed here (or at all). Can't we just lookup the module on `Base.__toplevel__`?

Reply:
> I'm nots ure loaded_path is needed here (or at all). Can't we just lookup the module on Base.__toplevel__?

no, it seems `Base.__toplevel__` is a quite special thing, it does not create any symbols inside.

Reply:
all the modules evaluated in it result as a root module.

Reply:
Alright, makes sense.

Reply:
I guess the same reason why there is a `Base.loaded_module` list